### ISSUE TITLE
[FIX-019] Represent empty graph validation/test splits as absent loaders

### ISSUE THREAD

**Issue by thekaveh** [thekaveh/NNx#229]
**Category:** Data contracts · **Priority:** P2 · **Size:** M

**Evidence status:** Reproduced dataset-integration defect.

## Outcome and scope

Represent empty graph validation/test masks by absent loaders in both sampler modes, and reject empty training masks at dataset construction. Keep full-graph context and seed-only scoring for nonempty splits. Do not add sampling backends, change graph splitting, or fold batch-size input validation from FIX-022 into this repair.

## Current behavior / evidence

`NNGraphDataset.__post_init__` computes batch sizes from mask sums, then constructs all three loaders unconditionally. Neighbor construction rejects a zero batch size; full mode constructs a graph whose `batch_size` is zero. `NNModel.train` treats any non-None validation loader as present. `NNDatasetBase` already permits optional val/test loaders, so this is a missing implementation of an existing contract.

## Repository map

- [`src/nnx/nn/dataset/nn_graph_dataset.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/dataset/nn_graph_dataset.py#L82) — Mask counts, mode dispatch, loader publication and displayed state.
- [`src/nnx/nn/dataset/nn_dataset_base.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/dataset/nn_dataset_base.py#L25) — Existing optional-loader types.
- [`src/nnx/nn/nn_model.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/nn_model.py#L1182) — Training validation gate; test the no-validation path.
- [`tests/test_nn_graph_dataset.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_nn_graph_dataset.py#L31) — Offline fake dataset and construction-only neighbor tests.
- [`tests/test_nn_graph_dataset_full_batch.py`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_nn_graph_dataset_full_batch.py#L82) — Permutation and seed-row behavior to preserve.

## Implementation approach

1. After loading the single graph, compute integer train/val/test mask counts once. Reject train count zero with a `ValueError` naming `train_mask`, before constructing loaders or publishing derived state. Do not mutate masks on the caller's graph.
2. Construct an optional loader only when that split count is positive. Set its field to `None` otherwise in both branches; in neighbor mode this also applies when an empty split has a positive requested size. Preserve full-mode rejection of every explicit size. Retain `shuffle`, generator and worker initialization arguments for nonempty neighbor loaders.
3. For an empty optional graph split, report resolved size zero and state string `"0"`; positive split resolution remains unchanged. Coordinate with FIX-022's validation of the original input tuple: zero derived from absence is legitimate, while explicitly requested zero is invalid.
4. Keep `_full_batch_loader` seed permutation and `input_id` semantics untouched for nonempty input. Exercise actual full-mode training in a temporary run root and assert the final record contains `val_edp is None`, rather than only testing dataset properties.

## Acceptance criteria

- [ ] Empty validation/test masks produce None in full and neighbor modes; nonempty split contents are unchanged.
- [ ] Full-batch model training with no validation split completes and records `val_edp=None`.
- [ ] An empty train mask produces an actionable construction-time ValueError.
- [ ] Existing permutation/seed-node scoring tests and nonempty neighbor-loader construction remain valid; no optional sampling dependency is needed for the full-mode tests.

## Validation and completion

Use this fully local dataset in the proposed regression:

```python
import torch
from torch_geometric.data import Data
from nnx import NNGraphDataset

class TinyGraph:
    num_features, num_classes = 2, 2
    def __init__(self, root, transform=None):
        self.data = Data(x=torch.eye(3, 2), y=torch.tensor([0, 1, 0]),
            edge_index=torch.tensor([[0, 1, 2], [1, 2, 0]]),
            train_mask=torch.tensor([True, True, False]),
            val_mask=torch.zeros(3, dtype=torch.bool),
            test_mask=torch.tensor([False, False, True]))
    def __getitem__(self, index):
        return self.data

ds = NNGraphDataset(ds_class=TinyGraph, sampler="full", n_workers=0)
assert ds.val_loader is None and ds.batch_sizes == (2, 0, 1)
assert ds.train_loader[0].batch_size == 2
```

Parameterize empty val, empty test, both empty and empty train. Neighbor cases only construct loaders and inspect masks; do not iterate them or require `pyg-lib`/`torch-sparse`.

Add `test_empty_optional_masks_return_none`, `test_empty_train_mask_rejected` and `test_full_graph_train_without_validation` to the two existing graph dataset test files. Run `uv run --frozen pytest tests/test_nn_graph_dataset.py tests/test_nn_graph_dataset_full_batch.py tests/test_graph_nn_base.py`. Preserve the existing full-batch parity test's local synthetic dataset; do not download Cora.

From the repository root, CI-equivalent setup is `uv sync --frozen --all-extras`. The commands above describe required implementation checks; none were executed during ticket enrichment. Finish with `uv run --frozen ruff check src/ tests/ examples/ scripts/`, `uv run --frozen ruff format --check src/ tests/ examples/ scripts/`, and `uv run --frozen pyright --warnings`. Preserve Python 3.10 compatibility and default-omission serialization; update canonical documentation sources, not generated projections by hand.

## Cascading changes and affected consumers

- **Optional-loader propagation.** [NNDatasetBase](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/dataset/nn_dataset_base.py#L25) already represents absent validation/test loaders with None. Pass those values unchanged into NNTrainParams and NNTrainerParams; neither loop should synthesize or iterate an empty validation loader. Both [NNModel validation dispatch](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/nn/nn_model.py#L1182) and [Trainer validation dispatch](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/src/nnx/trainer/trainer.py#L270) gate on None. Use an NNModel eval_step_fn spy and a Trainer model.evaluate spy to prove that neither validation route runs when the graph mask is empty, while ordinary training still commits LAST.
- **Metrics, callbacks and history.** A missing validation EDP must reach EarlyStopping, plateau fallback, progress display and saved IDPs as absence, not fabricated zero metrics. Use an explicit train-loss monitor when exercising EarlyStopping; its default validation monitor legitimately has no signal. Reload the run and retain `val_edp is None`. Keep the nonempty test split independently usable so absence of validation does not disable evaluation of seed rows.
- **Dataset metadata and example.** Preserve nonempty full-graph `input_id`/seed counts while reporting zero only for absent optional graph splits in state/table output. Requested batch-size validation remains [FIX-022](https://github.com/thekaveh/NNx/issues/232); full mode still rejects explicit sizes. Add an offline graph sample using a local PyG Data object, with no Planetoid/Cora download or neighbor iteration. The [sample discovery](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/tests/test_examples_smoke.py#L12) only finds numbered files: explicitly register the proposed unnumbered file for import and bounded execution, and list its PyG dependency (core in current metadata; use the graph extra only if [FEAT-031](https://github.com/thekaveh/NNx/issues/249) has landed) in the example index.

### Samples and executable coverage

- `examples/graph_optional_splits.py` (proposed): Proposed tiny synthetic PyG full-mode training example with nonempty train/test and empty validation; explicitly register import and execution discovery. **Verify:** Proposed test_graph_optional_splits_sample runs one epoch in tmp_path, asserts no eval-hook invocation, LAST exists and persisted validation stays absent.

Update `examples/README.md` with each affected sample’s purpose, extras, input/resource needs and bounded invocation. Preserve import-without-main behavior; add explicit import and execution registration in `tests/test_examples_smoke.py` for new unnumbered scripts. Existing discovery covers only `[0-9][0-9]_*.py`, and only selected scripts execute by default.

### Repository, site and wiki propagation

Edit these canonical sources; the two other columns are generated projections, not independently maintained pages. Proposed guide slugs below are explicit suggestions: reuse the prerequisite’s registered slug when it already exists.

| Canonical repository source | Generated site Markdown | Generated wiki Markdown | Required content change |
|---|---|---|---|
| [`docs/quickstart.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/docs/quickstart.md) | `generated/site/Quickstart.md` | `generated/wiki/Quickstart.md` | Add an offline full-graph example with an empty validation mask, val_loader=None and explicit train-loss monitoring. |
| [`examples/README.md`](https://github.com/thekaveh/NNx/blob/0c707eb5b80d2173843e60102ba326e4f997749d/examples/README.md) | `generated/site/Examples.md` | `generated/wiki/Examples.md` | Index the proposed optional-split graph sample and state that it uses synthetic data and no sampling backend. |

Update affected public docstrings/API catalog and regenerate `docs/api.md`; add the scoped changelog entry and reconcile changed top-level claims. Register new guides in `docs/manifest.yaml` with unique slugs/numbers and consistent headings/navigation. Repository-link validation uses Git-tracked paths, so newly linked source files must be tracked/staged before that check. Generated pages are never edited directly.

After source/API edits, run from the repository root in the existing frozen all-extras environment (regenerate/review the lock first only if this ticket changes dependencies):

```bash
uv run --frozen python -m scripts.docs.build_api_reference
uv run --frozen python -m scripts.docs.build_docs
uv run --frozen mkdocs build --strict
uv run --frozen python -m scripts.docs.build_wiki --check
uv run --frozen pytest tests/test_docs_projection.py tests/test_wiki_projection.py
```

If README changed, run `uv run --frozen python -m scripts.docs.build_pypi_readme` and its `--check` form. If canonical diagram HTML changed, run `uv run --frozen python -m scripts.docs.extract_architecture_svg` before the docs build and verify with `--check`. Compare the changed repository passages with rendered `site/` and wiki projections: values, signatures, limitations, examples, links/anchors and assets must agree. Deterministic generation alone is not a freshness or live-publication check.

**Delivery evidence:** record revision, executed checks and changed page mapping. `.github/workflows/docs.yml` builds PRs/main/develop and publishes Pages/wiki on main pushes. Verify served content after that authorized publication; record deployment failure/skew as pending follow-through. Local rendering is not live publication, and PR preparation requires no manual wiki/site push.

### Additional acceptance criteria — cascading completion

- [ ] Passing the absent graph validation loader through both training APIs invokes no evaluation hook and persists no fabricated validation metrics.
- [ ] Reloading the completed synthetic graph run preserves val_edp=None while evaluation of its nonempty test seed rows remains finite and sample counts remain correct.
- [ ] The explicitly registered offline graph sample executes without downloads or graph sampling extensions and displays zero for the absent split.
- [ ] The issue-specific changes in `docs/quickstart.md`, `examples/README.md`, affected API reference, sample catalog and changelog are reflected consistently in repository, generated/rendered site and wiki; documented commands/examples and links pass the checks above, and delivery evidence distinguishes the reviewed revision from any pending live publication.

**Coordination:** The traced downstream changes belong in this ticket. Shared-file coordination and optional combinations do not add a hard prerequisite.

Additional nonblocking consumers/coordination: [FIX-022](https://github.com/thekaveh/NNx/issues/232), [FIX-012](https://github.com/thekaveh/NNx/issues/222).

## Delivery order and dependencies

- **Execution order:** 30 / 67 (priority-aware topological order).
- **Dependency wave:** 1. Items in the same wave may proceed independently when their prerequisites are satisfied.
- **Blocked by:** None within this backlog.
- **Related / optional integration:** [FIX-012](https://github.com/thekaveh/NNx/issues/222), [FIX-022](https://github.com/thekaveh/NNx/issues/232), [FEAT-026](https://github.com/thekaveh/NNx/issues/246). These links do not block this ticket.

**Dependency rationale.** Independent repair of current behavior. Coordinate with FIX-022 for batch-size validation; empty optional masks have separate semantics.

Reviewed against [`0c707eb`](https://github.com/thekaveh/NNx/commit/0c707eb5b80d2173843e60102ba326e4f997749d) on 2026-09-20. Acceptance criteria are prospective; this issue does not claim an implementation or new test pass.

**Comment by thekaveh** [comment]
Resolved in #303 (develop, merged `6d15332`) and #304 (main, merged `8afa774`).

**Root cause.** `NNGraphDataset.__post_init__` derived each split's batch size from its mask sum and built all three loaders unconditionally. With an empty `val_mask`, full mode produced a one-element list holding a zero-node batch that `NNModel.train` (which treats any non-`None` loader as present) fed to `evaluate()` — failing with `evaluate() loader produced zero samples`; neighbor mode could not construct at all (`batch_size should be a positive integer value, but got batch_size=0`); and an empty `train_mask` was silently accepted as a zero-size training loader.

**What changed** (`src/nnx/nn/dataset/nn_graph_dataset.py` only — the training gates already honour `None`):
- The three mask counts are read once (the caller's graph is never edited); an empty `train_mask` raises a `ValueError` naming it before any loader or derived state exists.
- An empty `val_mask` / `test_mask` is an absent split in both sampler modes: that loader is `None`, its resolved `batch_sizes` entry is `0` (derived from absence; FIX-022's validator still rejects an explicit zero request) and `state()` reports `"0"`. In neighbor mode an explicit positive request for an empty split also yields `None`. Nonempty neighbor loaders keep `shuffle`, the seeded generator and `worker_init_fn`; nonempty full-batch loaders keep the permutation / `input_id` seed-row semantics; `sampler="full"` still rejects explicit sizes.
- New offline `examples/graph_optional_splits.py` (six-node in-memory PyG graph, empty validation mask, full mode, `eval_step_fn` spy, `EarlyStopping(monitor="train_edp.loss")`, LAST + `NNRun.load` round trip, test split scored on its seed rows), registered explicitly in `tests/test_examples_smoke.py` for import and bounded execution. Docs: `NNGraphDataset` docstring, `docs/quickstart.md` §2.12, `examples/README.md`, `CHANGELOG.md [Unreleased]`, regenerated `docs/api.md`.

**Acceptance criteria → tests:**
- Empty val / test / both masks → `None` in full and neighbor modes, nonempty splits unchanged, caller masks untouched, explicit request on an empty neighbor split → `None` — `test_empty_optional_masks_return_none` (`tests/test_nn_graph_dataset.py`, 6 cases).
- Empty train mask → construction-time `ValueError` naming `train_mask` — `test_empty_train_mask_rejected` (both samplers).
- Full-batch training with no validation split: no evaluation hook runs (`eval_step_fn` spy on `NNModel.train`, `model.evaluate` spy on `Trainer.train`), every IDP has `val_edp=None` — also after `NNRun.load` — LAST committed, `EarlyStopping` on a train monitor works, test split evaluated finitely on its 3 seed rows — `test_full_graph_train_without_validation` (`tests/test_nn_graph_dataset_full_batch.py`).
- Existing permutation / seed-node / neighbor-construction tests unchanged and green; full-mode tests need no sampling backend.
- Registered offline sample runs without downloads or sampling extensions and displays `"0"` for the absent split — `graph_optional_splits_workflow`.

**Checks (both PRs, local + CI green: lint-and-test 3.10–3.14, pip-audit, CodeQL, docs build):**
```
uv run --frozen pytest -q                                    → 2143 passed, 1 skipped, 2 deselected (libcairo docs-projection tests, CI runs them)
uv run --frozen ruff check src/ tests/ examples/ scripts/    → All checks passed
uv run --frozen ruff format --check …                        → 258 files already formatted
uv run --frozen pyright --warnings                           → 0 errors, 0 warnings
build_api_reference / build_docs / mkdocs build --strict / build_wiki --check / projection tests → pass
python examples/graph_optional_splits.py                     → batch_sizes=(3, 0, 3), val_edp_persisted=[None, None], test_seed_rows=3
```
Not verifiable locally: neighbor-mode loader iteration (needs pyg-lib / torch-sparse); neighbor cases construct and inspect only. Live site/wiki publication runs from the main push; not re-verified here.
